stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp

```diff
@@ -0,0 +1,206 @@
+#ifndef STAN_MATH_PRIM_MAT_PROB_CATEGORICAL_LOGIT_GLM_LPMF_HPP
+#define STAN_MATH_PRIM_MAT_PROB_CATEGORICAL_LOGIT_GLM_LPMF_HPP
+
+#include <stan/math/prim/mat/meta/as_column_vector_or_scalar.hpp>
+#include <stan/math/prim/scal/err/check_bounded.hpp>
+#include <stan/math/prim/scal/err/check_consistent_size.hpp>
+#include <stan/math/prim/scal/fun/size_zero.hpp>
+#include <stan/math/prim/scal/meta/partials_return_type.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/meta/operands_and_partials.hpp>
+#include <stan/math/prim/scal/meta/as_array_or_scalar.hpp>
+#include <stan/math/prim/scal/meta/include_summand.hpp>
+#include <stan/math/prim/scal/meta/is_constant_struct.hpp>
+#include <stan/math/prim/scal/meta/scalar_seq_view.hpp>
+#include <Eigen/Core>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the log PMF of the Generalized Linear Model (GLM)
+ * with categorical distribution and logit (softmax) link function.
+ *
+ * @tparam T_y type of classes. It can be either `std::vector<int>` or `int`.
+ * @tparam T_x_scalar type of a scalar in the matrix of independent variables
+ * (features)
+ * @tparam T_x_rows compile-time number of rows of `x`. It can be either
+ * `Eigen::Dynamic` or 1.
+ * @tparam T_alpha_scalar type of scalar in the intercept vector
+ * @tparam T_beta_scalar type of a scalar in the matrix of weights
+ * @param y a scalar or vector of classes. If it is a scalar it will be
+ * broadcast - used for all instances. Values should be between 1 and number of
+ * classes, including endpoints.
+ * @param x design matrix or row vector. If it is a row vector it will be
+ * broadcast - used for all instances.
+ * @param alpha intercept vector (in log odds)
+ * @param beta weight matrix
+ * @return log probability or log sum of probabilities
+ * @throw std::domain_error x, beta or alpha is infinite or y is not within
+ * bounds
+ * @throw std::invalid_argument if container sizes mismatch.
+ */
+template <bool propto, typename T_y, typename T_x_scalar, int T_x_rows,
+          typename T_alpha_scalar, typename T_beta_scalar>
+typename return_type<T_x_scalar, T_alpha_scalar, T_beta_scalar>::type
```

Comment:
This might be a lot to ask, but is there anyway we can make the templating more readable like @SteveBronder did in the gpu_cov_exp_quad?

Is it possible to write a meta program that is a general return type for all of our pmf/pdfs? We could just make this a default moving forward, reading templates were a huge issue for me starting to learn this library.

Reply:
I can write something similar so that we have 
`typename return_type_t<T_x_scalar, T_alpha_scalar, T_beta_scalar>` 
instead of 
`typename return_type<T_x_scalar, T_alpha_scalar, T_beta_scalar>::type`

Reply:
I'd say let's pass because it's not at all helping with readability/reducing lines of code.